### PR TITLE
Restructured and updated project

### PR DESCRIPTION
**Pull request by Atrejoe**
Made solution with separated examples and unit tests
Replaced VS Test with Xunit test
Upgraded to .Net Framework 4.8

### CODE REVIEW

**Comment by AaronAtDuo** [comment]
@Atrejoe Sorry for getting to this so late.  I'm definitely looking to clean up and modernize this repo a bit; it's been neglected for a while.  Do yo think it would be worthwhile migrating off .NET Framework entirely, and targeting .NET Standard (like duo_universal_csharp)?

**Comment by Atrejoe** [comment]
For libraries in general: certainly. This should not be a big chore though.

I do not have my current code with me, but I think I have switched to:
https://github.com/duosecurity/duo_universal_csharp

**Comment by markcouvaras** [comment]
Hi there, sorry to ask this on here but I was wondering if anyone knew how to call JSONPagingApiCall using 'v2/logs/authentication' ? I don't know how to use the offset parameter as it is no longer an int. Would be nice to update the examples file to show to use in the V2 api?

If anyone can help, it would greatly appreciated.

File: duo_api_csharp/duo_api_csharp.csproj [hunk 1]
```diff
@@ -1,89 +1,55 @@
...
+    <Reference Include="System.Data" />
+    <Reference Include="System.Data.DataSetExtensions" />
+    <Reference Include="System.Web" />
+    <Reference Include="System.Web.Extensions" />
+    <Reference Include="System.Xml" />
+    <Reference Include="System.Xml.Linq" />
+  </ItemGroup>
+  <ItemGroup>
+    <Compile Include="Duo.cs" />
+    <Compile Include="DuoApiMethods.cs" />
```
Comment (yizshi) [inline duo_api_csharp/duo_api_csharp.csproj]: Sorry to ask obvious question, but I could not find this file in your diff. Maybe the `DuoApiMethods.cs` was left in your commit.

**Review by yizshi — CHANGES_REQUESTED** [review]
Thank you so much on working on this! I really love all the changes you have made, and it looks like going to make code change and testing in the future so much easier.

One quick note, we added github action in master couple weeks ago. It would be awesome if you could rebase with master branch so we can test it :) . I think there are might be some conflict due to some change on the signing test. The file is recently modified. It _should_ be fairly simple.

I do have a comment left in the file. Also couple question about rename/deleting file. I could not find a way to comment on the renamed file on github PR. So I am going to ask here :) 
I notice you delete the `DuoAPiTest.sln`, sadly, that file was used on github action. But I do think we can modify ci yml file to make it work.
Another thing I notice is you moved the `Makefile`. Looked on the content, make file are assume running inside `test` folder. Although I do not think we need to use `Makefile` anymore. 
I am not that familiar with dotnet, but I assume with this upgrade, you can run the test with `dotnet` command? If that's the case, I think we could just remove the `Makefile`

**Comment by Atrejoe** [comment]
I rebased the branch.

indeed DuoApiMethods.c was part of my other feature branch.
When you look at our copy, you'll also see a port to .Net Standard & XUnit, and a branch to make the library more user friendly by having convenience methods for calling the DUO api. (see `DuoApiMethods.cs`)

image

**Comment by yizshi** [comment]
Hi - @Atrejoe 
Thank you for rebasing! There are just couple things I want to ask before I merge this.
1. In the attempt I tried to run this locally as well as in github action. I notice `VS2019` as well as `dotnet` could not find the tests. I have it working after I added `xunit.runner.visualstudio.2.4.5` from nuget. I think it might be worth to add it in this diff too.
2. After I got the test running. I notice one test could not pass. `TestApiCall.TestJsonUnparseableResponse`. I am suspect this is a because the unit tests are try to listen to a losthost that already in use for another test. I am still looking into how to get it fix. But if you could get it fix before me, that would fantastic :P.

**Comment by Atrejoe** [comment]
I have added `xunit.runner.visualstudio`, I didn't need it in VS2022, but dotnet did.
Please note that this is currently still a .Net framework library. It may worth it to port the library to .Net standard and then iron out these issues.

This issue with `TestApiCall.TestJsonUnparseableResponse` may be a concurrency issue.
Multiple tests use the same instance on the TestServer. When run concurrently these tests they change the handler then perform the tests. In concurrent scenario's it may be that the handler get changed before performing the actual test.

I have another issue with `TestPostParameterSigningCustomDate` which indicates `Bad Date!`.
This may be due to unexpectyed time zone handling:

Expected value:
`Mon, 11 Nov 2013 22:34:00 +0000`
my actual value:
`Mon, 11 Nov 2013 22:34:00 +0100`

A UTC dat is passed in as a parameter, but then in `DateToRFC822` the offset is determined using `TimeZoneInfo.Local`, this may need to be `TimeZoneInfo.UTC` dependant on the type of time passed in:

```
            TimeSpan offsetTime= default;
            switch (date.Kind)
            {
                case DateTimeKind.Unspecified:
                    //Don't know yet
                    break;
                case DateTimeKind.Utc:
                    {
                        offsetTime = TimeZoneInfo.Utc.GetUtcOffset(date);
                        break;
                    }
                case DateTimeKind.Local:
                    {
                        offsetTime = TimeZoneInfo.Local.GetUtcOffset(date);
                        break;
                    }
            };

            int offset = offsetTime.Hours;
```

I haven't looked at the purpose of the method, so I do not know for sure though.

**Comment by AaronAtDuo** [comment]
FWIW we do (long term) want to port it to .NET standard.  We have a short-term need to get to .NET 4.8 framework ASAP though, so we're working on getting this merged.  We really appreciate you working on this with us.

**Review by yizshi — APPROVED** [review]
This overall looks good to me. 
I run the test locally. There are two tests failing currently
`TestJsonUnparseableResponse` -> This test always run after `TestJsonTimeout` and because we use `thread.sleep`. The test http server can not close in timely fashion. This is not relevant to this PR because this is just a bad designed test. I am going to look into fix it once it merged to master.
`TjestPostParameterSigningCustomDate` -> The data set with `DateTime` did not honor the `DateTimeKind.Utc` setting. I have not look too deep. But again this is not relevant to this PR.

@Atrejoe Thank you again to make this upgrade. It is not a easy task.

**Pull request merged**